### PR TITLE
docs: GEP 1709 profiles at the *Route granularity level

### PR DESCRIPTION
**Pull request by shaneutt**
**What type of PR is this?**

/kind gep
/kind documentation

**What this PR does / why we need it**:

There's been a lot of feedback from the community in community meetings and on recent PRs that there's a preference for named profiles to correspond directly with `*Route` APIs, and move down a level in granularity from `Layer 4` and `Layer 7`. This PR makes the language changes necessary to support that desire.

### CODE REVIEW

**Comment by youngnick** [comment]
/hold for multiple reviews

but
/lgtm

File: geps/gep-1709.md
Comment (sunjayBhatia) [inline geps/gep-1709.md:113]: where would the existing tests corresponding to Gateway/GatewayClass validation/features fall here? would they also automatically be run?

**Review by sunjayBhatia — COMMENTED** [review]
ultimately for Contour I don't think we're too picky about what the different conformance profiles are (so I don't hold too strong of an opinion on what we go with), but some thoughts/ideas that might spark discussion and some since people are enthusiastic about breaking up the profiles more granularly:

For implementers, having a little more granularity makes sense
  - Allows you to possibly pick and choose functionality to implement iteratively while not having to monitor every test that is added
  - Easy to see in a report what is passing/failing/skipped etc.

If we intend to use the conformance reports to present Gateway API implementations (and the API itself) to users, might it make more sense to break up the higher level (e.g. L7) groupings of tests along functionality lines rather than by resource?
- at least at first this might be more of a naming distinction tbh, since I think the intent of the per-*Route profiles is pretty similar
- e.g. subset of L7 tests that correspond to HTTP (basically HTTPRoute tests, maybe some TLSRoute tests for HTTP over TLS passthrough)
- e.g. subset of L7 tests that correspond to gRPC (basically GRPCRoute tests)
- as a user I might not know/care which resource I need to use for what purpose, especially when coming from Ingress v1, so evaluating implementation's conformance with that as the first "filter" is less useful?
- as someone who is totally new to Gateway API, if I see the available features expressed as behavior/functionality that might be more compelling to migrate to Gateway API?

I guess the above ^ is related to the idea of nailing down what the target audience for the conformance report is, if it is not just Gateway API maintainers/implementers, we might want to target how we present things (or decide that work is done in the website UX presentation of conformance reports, rather than the raw report format)

**Comment by shaneutt** [comment]
> 

So it sounds like you're advocating for the possibility that we might _want_ the higher level ones eventually, which is a thought I share. Currently @sunjayBhatia I've placed this in the `alternatives considered` which is like the `TODO` area in that it's stuff to follow up on later, but unlike `TODO` this change would _not_ consider it a blocker for this KEP to move to `implementable`. As I see it, there's nothing stopping us from adding higher levels of composition at a later time given what we have so far, does that ring true with you as well? Basically: are there any specific actions you want to see taken here and now, as a part of this PR before you would be willing to approve this PR?

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:113]: That's a good point. Implicitly we understand they would run under the route named profiles, but we didn't actually explicitly spell that out yet. f9d04b965ece6074d52ae3988f8fb4cc704c1e53 takes care of adding a note about this :+1:

**Comment by sunjayBhatia** [comment]
> > 
> 
> So it sounds like you're advocating for the possibility that we might _want_ the higher level ones eventually, which is a thought I share. Currently @sunjayBhatia I've placed this in the `alternatives considered` which is like the `TODO` area in that it's stuff to follow up on later, but unlike `TODO` this change would _not_ consider it a blocker for this KEP to move to `implementable`. As I see it, there's nothing stopping us from adding higher levels of composition at a later time given what we have so far, does that ring true with you as well? Basically: are there any specific actions you want to see taken here and now, as a part of this PR before you would be willing to approve this PR?

I think that TODO section is clear about wanting to leave open the possibility for "higher level" profiles 👍🏽, but my comment is more about seeing what people think about reframing how the granular breakdown of profiles is presented, i.e. from resource kind-centric to functional area/feature-centric (e.g. "HTTP" rather than "HTTPRoute") since new users of GW API might not know/care about what resources they need to use, just what functionality they want to mirror from Ingress (or some implementation-specific CR) for example

**Comment by shaneutt** [comment]
> > > 
> > 
> > 
> > So it sounds like you're advocating for the possibility that we might _want_ the higher level ones eventually, which is a thought I share. Currently @sunjayBhatia I've placed this in the `alternatives considered` which is like the `TODO` area in that it's stuff to follow up on later, but unlike `TODO` this change would _not_ consider it a blocker for this KEP to move to `implementable`. As I see it, there's nothing stopping us from adding higher levels of composition at a later time given what we have so far, does that ring true with you as well? Basically: are there any specific actions you want to see taken here and now, as a part of this PR before you would be willing to approve this PR?
> 
> I think that TODO section is clear about wanting to leave open the possibility for "higher level" profiles 👍🏽, but my comment is more about seeing what people think about reframing how the granular breakdown of profiles is presented, i.e. from resource kind-centric to functional area/feature-centric (e.g. "HTTP" rather than "HTTPRoute") since new users of GW API might not know/care about what resources they need to use, just what functionality they want to mirror from Ingress (or some implementation-specific CR) for example

Ah I was misunderstanding before. Now I get what you're saying. That sounds good to me actually, let's try that out. 0526ef85d59aae73e8eb5e23f3d5dc89a7e2e1cb makes the change :+1:

**Review by sunjayBhatia — APPROVED** [review]
I'm on board with this change, interested in what others think!

File: geps/gep-1709.md
Comment (arkodg) [inline geps/gep-1709.md:118]:
```suggestion
- `GRPC`
```

File: geps/gep-1709.md
Comment (arkodg) [inline geps/gep-1709.md:111]: does this cover passthrough or termination ?

Comment (robscott) [reply geps/gep-1709.md:111]: This is a good question, it's probably safest to just name these profiles `HTTPRoute`, `TLSRoute`, etc, to avoid any confusion with overall protocol support.

**Review by robscott — COMMENTED** [review]
Thanks @shaneutt! This mostly LGTM.

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:111]: @sunjayBhatia any thoughts here?

**Comment by shaneutt** [comment]
Adding a hold on merge so that I can do some squashing and a rebase prior to merge.

File: geps/gep-1709.md
Comment (sunjayBhatia) [reply geps/gep-1709.md:111]: thinking about the `TLS` category name a little, it is a bit broad, you can do HTTP over TLS, gRPC over TLS, some other TCP protocol etc., and then you get the terminate/passthrough configuration option

I think as-written the `TLS` category is intended to cover [this existing test](https://github.com/kubernetes-sigs/gateway-api/blob/40a86dcad8f2dd307435b82317483248d484f853/conformance/tests/tlsroute-simple-same-namespace.go) and any other related additional tests. Given our stance on listener protocol/route type (https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls), TLSRoute is only really used for passthrough TLS, so conceptually I think we can express that.

IMO we could rename `TLS` here to `TLSPassthrough`

Comment (shaneutt) [reply geps/gep-1709.md:111]: Let's go for that, again we're still in provisional/prototyping so we've got lots of runway to try this out and see how it feels.

```suggestion
- `TLSPassthrough`
```

Please anyone unresolve if you feel that there's more action to take here :+1:

Comment (shaneutt) [reply geps/gep-1709.md]: `ReferenceGrant` is actually one that's been troubling me. We have a note about how `GatewayClass` and `Gateway` will be automatically included as they are referent in some featuresets, see here:

https://github.com/kubernetes-sigs/gateway-api/pull/1783/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R126

However I only put `GatewayClass` and `Gateway` there because those are easy and obvious, `ReferenceGrant` _might_ fit there but it's a little more subtle.

Since we'll need to make sure we account for this before we go GA, but I would prefer not to get this PR hung up for it I've added an explicit TODO item saying that we'll need to resolve this before `implementable`:

https://github.com/kubernetes-sigs/gateway-api/pull/1783/commits/d7bd41733e5f6ce40bbd45cb7f09ddac0863a71c

**Review by youngnick — COMMENTED** [review]
I think this LGTM for provisional, given we've got a bunch of TODOs still.

/lgtm
/unhold

**Pull request merged**
